modules/xmpp/SignalingLayerImpl.js

```diff
@@ -51,40 +98,211 @@ export default class SignalingLayerImpl extends SignalingLayer {
+            }
         }
     }
 
+    /**
+     * Finds the first source of given media type for the given endpoint.
+     * @param endpointId
+     * @param mediaType
+     * @returns {SourceInfo|null}
+     * @private
+     */
+    _findEndpointSourceInfoForMediaType(endpointId, mediaType) {
+        const remoteSourceState = this._remoteSourceState[endpointId];
+
+        if (!remoteSourceState) {
+            return null;
+        }
+
+        for (const sourceInfo of Object.values(remoteSourceState)) {
+            const _mediaType = getMediaTypeFromSourceName(sourceInfo.sourceName);
+
+            if (_mediaType === mediaType) {
+                return sourceInfo;
+            }
+        }
+
+        return null;
+    }
+
     /**
      * @inheritDoc
      */
     getPeerMediaInfo(owner, mediaType) {
-        if (this.chatRoom) {
-            return this.chatRoom.getMediaPresenceInfo(owner, mediaType);
+        const legacyGetPeerMediaInfo = () => {
+            if (this.chatRoom) {
+                return this.chatRoom.getMediaPresenceInfo(owner, mediaType);
+            }
+            logger.error('Requested peer media info, before room was set');
+        };
+
+        if (FeatureFlags.isSourceNameSignalingEnabled()) {
+            const lastPresence = this.chatRoom.getLastPresence(owner);
+
+            if (!lastPresence) {
+                throw new Error(`getPeerMediaInfo - no presence stored for: ${owner}`);
+            }
+
+            if (!this._doesEndpointSendNewSourceInfo(owner)) {
+                return legacyGetPeerMediaInfo();
+            }
+
+            /**
+             * @type {PeerMediaInfo}
+             */
+            const mediaInfo = {};
+            const endpointMediaSource = this._findEndpointSourceInfoForMediaType(owner, mediaType);
+
+            // The defaults are provided only, because getPeerMediaInfo is a legacy method. This will be eventually
+            // changed into a getSourceInfo method which returns undefined if there's no source. Also there will be
+            // no mediaType argument there.
+            if (mediaType === MediaType.AUDIO) {
+                mediaInfo.muted = endpointMediaSource ? endpointMediaSource.muted : true;
+            } else if (mediaType === MediaType.VIDEO) {
+                mediaInfo.muted = endpointMediaSource ? endpointMediaSource.muted : true;
+                mediaInfo.videoType = endpointMediaSource ? endpointMediaSource.videoType : undefined;
+
+                const codecTypeNode = filterNodeFromPresenceJSON(lastPresence, 'jitsi_participant_codecType');
+
+                if (codecTypeNode.length > 0) {
+                    mediaInfo.codecType = codecTypeNode[0].value;
+                }
+            } else {
+                throw new Error(`Unsupported media type: ${mediaType}`);
+            }
+
+            return mediaInfo;
         }
-        logger.error('Requested peer media info, before room was set');
+
+        return legacyGetPeerMediaInfo();
```

Comment:
if the feature flag is disabled, the legacy handling was left unchanged, but moved into a method, so can be re-used as part of the new flow (temporarily)